File: src/three_sum_closest/three_sum_closest.py
#!/usr/bin/env python
# coding:utf-8
# filename:three_sum_closest.py

'''
author:	gavingeng
date:	2018-06-10 10:12:34
'''


class Solution(object):
    def threeSumClosest(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        nums = sorted(nums)
        ret = 0
        minGap = 10 ** 10

        for i in range(len(nums)):
            j = i + 1
            k = len(nums) - 1

            while (j < k):
                sum = nums[i] + nums[j] + nums[k]
                gap = abs(sum - target)
                if gap < minGap:
                    ret = sum
                    minGap = gap

                if sum < target:
                    j += 1
                else:
                    k -= 1

        return ret


def test():
    s = Solution()
    nums = [-1, 2, 1, -4]
    target = 1
    ret = s.threeSumClosest(nums, target)
    print ret


def main():
    test()


if __name__ == '__main__':
    main()

File: src/three_sum_closest/ThreeSumClosest.java
package three_sum_closest;

import java.util.Arrays;

/**
 * Created by gavingeng on 2018-06-10 10:12.
 */
public class ThreeSumClosest {
	
	/**
	 * 排序，然后左右夹逼
	 * 比3sum多引入一个间隔minGap，用这个来判断接近程度
	 *
	 * @param nums
	 * @param target
	 * @return
	 */
	public static int threeSumClosest(int[] nums, int target) {
		Arrays.sort(nums);
		
		int ret = 0;
		int minGap = Integer.MAX_VALUE;
		for (int i = 0; i < nums.length - 1; i++) {
			int j = i + 1;
			int k = nums.length - 1;
			
			while (j < k) {
				int sum = nums[i] + nums[j] + nums[k];
				int gap = Math.abs(sum - target);
				if (gap < minGap) {
					ret = sum;
					minGap = gap;
				}
				
				if (sum < target) {
					j++;
				} else {
					k--;
				}
			}
		}
		return ret;
	}
	
	public static void main(String[] args) {
//		int[] nums = {-1, 2, 1, -4};
//		int target = 1;
		int[] nums = {1, 1, 1, 0};
		int target = 2;
		int ret = threeSumClosest(nums, target);
		System.out.println(ret);
	}
}
